File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.kt
class Solution {
    fun numOfMinutes(n: Int, headID: Int, manager: IntArray, informTime: IntArray): Int {
        val finallyInformedTime = mutableListOf<Int>()
        val employees = HashMap<Int, MutableList<Int>>()

        for (i in 0 until n) {
            if (manager[i] == -1) continue
            employees[manager[i]]?.let { employees[manager[i]]!!.add(i) }
                ?: employees.put(manager[i], mutableListOf(i))
        }

        informToSubordinates(0, headID, informTime, employees, finallyInformedTime)
        return finallyInformedTime.max() ?: 0
    }

    private fun informToSubordinates(time: Int, currentEmployee: Int, informTime: IntArray, employees: HashMap<Int, MutableList<Int>>, finallyInformedTime: MutableList<Int>) {
        if (!employees.containsKey(currentEmployee) || employees[currentEmployee]!!.isEmpty()) {
            finallyInformedTime.add(time)
            return
        }
        val nextTime = time + informTime[currentEmployee]
        for (next in employees[currentEmployee]!!) {
            informToSubordinates(nextTime, next, informTime, employees, finallyInformedTime)
        }
    }
}

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.cpp
class Solution {
private:
    void informToSubordinates(int time, int currentEmployee, vector<int> &informTime, unordered_map<int, vector<int>> &employees, vector<int> &finallyInformedTimes) {
        if (currentEmployee == -1 || employees[currentEmployee].empty()) {
            finallyInformedTimes.push_back(time);
            return;
        }

        int nextTime = time + informTime[currentEmployee];
        for (auto &next : employees[currentEmployee]) {
            informToSubordinates(nextTime, next, informTime, employees, finallyInformedTimes);
        }
    }

public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        unordered_map<int, vector<int>> employees;
        vector<int> finallyInformedTimes;

        for (int i = 0; i < n; i++) {
            employees[manager[i]].push_back(i);
        }

        informToSubordinates(0, headID, informTime, employees, finallyInformedTimes);
        return *max_element(finallyInformedTimes.begin(), finallyInformedTimes.end());
    }
};

